File: src/cs/Bootsharp.Common/Interop/Instances.cs
namespace Bootsharp;

/// <summary>
/// Manages exported (C# -> JavaScript) instanced interop interfaces.
/// </summary>
public static class Instances
{
    private static readonly Dictionary<int, object> idToInstance = [];
    private static readonly Queue<int> idPool = [];
    private static int nextId = int.MinValue;

    /// <summary>
    /// Registers specified interop instance and associates it with unique ID.
    /// </summary>
    /// <param name="instance">The instance to register.</param>
    /// <returns>Unique ID associated with the registered instance.</returns>
    public static int Register (object instance)
    {
        var id = idPool.Count > 0 ? idPool.Dequeue() : nextId++;
        idToInstance[id] = instance;
        return id;
    }

    /// <summary>
    /// Resolves registered instance by the specified ID.
    /// </summary>
    /// <param name="id">Unique ID of the instance to resolve.</param>
    public static object Get (int id)
    {
        if (!idToInstance.TryGetValue(id, out var instance))
            throw new Error($"Failed to resolve exported interop instance with '{id}' ID: not registered.");
        return instance;
    }

    /// <summary>
    /// Notifies that interop instance is no longer used on JavaScript side
    /// (eg, was garbage collected) and can be released on C# side as well.
    /// </summary>
    /// <param name="id">ID of the disposed interop instance.</param>
    public static void Dispose (int id)
    {
        if (!idToInstance.Remove(id))
            throw new Error($"Failed to dispose exported interop instance with '{id}' ID: not registered.");
        idPool.Enqueue(id);
    }
}

File: src/js/src/instances.ts
import { exports } from "./exports";

const exportedFinalizer = new FinalizationRegistry(finalizeExported);
const exportedById = new Map<number, WeakRef<object>>();
const importedById = new Map<number, object>();
const idByImported = new Map<object, number>();
const onDisposeById = new Map<number, () => void>();
const idPool = new Array<number>();
let nextId = -2147483648; // Number.MIN_SAFE_INTEGER is below C#'s Int32.MinValue

export const instances = {
    /** Invokes the specified factory to create and register an exported instance wrapper associated with the ID,
     *  unless an exported instance is already registered under the ID, in which case returns its wrapper. */
    export(id: number, factory: (id: number) => object): object {
        const instance = exportedById.get(id)?.deref();
        if (instance != null) return instance;
        const exported = factory(id);
        exportedById.set(id, new WeakRef(exported));
        exportedFinalizer.register(exported, id);
        return exported;
    },
    /** Registers specified imported instance and associates it with a unique ID, unless it's already registered,
     *  in which case the ID of the registered instance is returned. */
    import(instance: object, factory?: (id: number) => () => void): number {
        const registered = idByImported.get(instance);
        if (registered !== undefined) return registered;
        const id = idPool.length > 0 ? idPool.pop()! : nextId++;
        importedById.set(id, instance);
        idByImported.set(instance, id);
        if (factory != null) onDisposeById.set(id, factory(id));
        return id;
    },
    /** Returns a registered imported instance associated with the specified ID. */
    imported(id: number): object {
        return importedById.get(id)!;
    },
    /** Invoked from C# to notify that the imported (JS -> C#) instance is no longer used
     *  (eg, was garbage collected) and can be released on the JavaScript side as well.
     *  @param id Unique identifier of the disposed instance. */
    disposeImported(id: number): void {
        idByImported.delete(importedById.get(id)!);
        importedById.delete(id);
        onDisposeById.get(id)?.();
        onDisposeById.delete(id);
        idPool.push(id);
    }
};

/* v8 ignore start -- @preserve */ // Uncoverable, as finalization in Node is not controllable.
function finalizeExported(id: number) {
    exportedById.delete(id);
    (<{ DisposeExportedInstance: (id: number) => void }>exports).DisposeExportedInstance(id);
}
/* v8 ignore stop -- @preserve */
